### PR TITLE
Add 'promotion' methods to Sample classes.

### PR DESCRIPTION
**Pull request by olliethomas**

### CODE REVIEW

File: bolts/data/structures.py [hunk 1]
```diff
@@ -31,27 +31,70 @@ class NamedSample:
     def __len__(self) -> int:
         return len(self.__dataclass_fields__)  # type: ignore[attr-defined]
 
+    def add_field(
+        self, y: Tensor | None = None, s: Tensor | None = None, iw: Tensor | None = None
+    ) -> NamedSample | BinarySample | BinarySampleIW | TernarySample | TernarySampleIW:
```
Comment (MylesBartlett) [inline bolts/data/structures.py:36]: overrrrload

File: bolts/data/structures.py [hunk 2]
```diff
@@ -25,32 +25,131 @@
...
+    @overload
+    def add_field(self, *, y: Tensor, s: Tensor) -> TernarySample:
+        ...
+
+    @overload
+    def add_field(self, *, y: Tensor, s: Tensor, iw: Tensor) -> TernarySampleIW:
+        ...
+
+    def add_field(
+        self, y: Tensor | None = None, s: Tensor | None = None, iw: Tensor | None = None
```
Comment (MylesBartlett) [inline bolts/data/structures.py:59]:
```suggestion
        self, *, y: Tensor | None = None, s: Tensor | None = None, iw: Tensor | None = None
```

File: bolts/data/structures.py [hunk 3]
```diff
@@ -25,32 +25,131 @@
...
+    @overload
+    def add_field(self, s: Tensor | None = None) -> TernarySample:
+        ...
+
+    @overload
+    def add_field(self, s: Tensor | None = None, iw: Tensor | None = None) -> TernarySampleIW:
+        ...
+
+    def add_field(
+        self, s: Tensor | None = None, iw: Tensor | None = None
```
Comment (MylesBartlett) [inline bolts/data/structures.py:96]:
```suggestion
        self, *, s: Tensor | None = None, iw: Tensor | None = None
```

File: bolts/data/structures.py [hunk 4]
```diff
@@ -25,32 +25,131 @@
...
+    @overload
+    def add_field(self, iw: Tensor | None = None) -> BinarySampleIW:
+        ...
+
+    @overload
+    def add_field(self, s: Tensor | None = None) -> TernarySample:
+        ...
+
+    @overload
+    def add_field(self, s: Tensor | None = None, iw: Tensor | None = None) -> TernarySampleIW:
```
Comment (MylesBartlett) [inline bolts/data/structures.py:92]:
```suggestion
    def add_field(self, *, s: Tensor | None = None, iw: Tensor | None = None) -> TernarySampleIW:
```

File: bolts/data/structures.py [hunk 5]
```diff
@@ -25,32 +25,131 @@
...
+    @overload
+    def add_field(self) -> BinarySample:
+        ...
+
+    @overload
+    def add_field(self, iw: Tensor | None = None) -> BinarySampleIW:
+        ...
+
+    @overload
+    def add_field(self, s: Tensor | None = None) -> TernarySample:
```
Comment (MylesBartlett) [inline bolts/data/structures.py:88]:
```suggestion
    def add_field(self, *, s: Tensor | None = None) -> TernarySample:
```

File: bolts/data/structures.py [hunk 6]
```diff
@@ -25,32 +25,131 @@
...
 
 @dataclass(frozen=True)
-class BinarySampleIW(BinarySample):
+class BinarySample(_SampleBase, _BinarySampleMixin):
+    @overload
+    def add_field(self) -> BinarySample:
+        ...
+
+    @overload
+    def add_field(self, iw: Tensor | None = None) -> BinarySampleIW:
```
Comment (MylesBartlett) [inline bolts/data/structures.py:84]:
```suggestion
    def add_field(self, *, iw: Tensor | None = None) -> BinarySampleIW:
```

File: tests/data_test.py [hunk 7]
```diff
@@ -18,3 +26,29 @@ def test_datasets(ds_cls: type[VisionDataset]) -> None:
     ds = ds_cls(root=Path("~/Data").expanduser(), transform=transform)
     for _ds in ds:
         assert _ds[0] is not None
+
+
+def test_add_field():
```
Comment (olliethomas) [inline tests/data_test.py:31]:
```suggestion
def test_add_field() -> None:
```

**Pull request merged**